File: infrastructure/ts-sdk/src/platform.ts
/**
 * Unified platform client holding all 12 infrastructure component clients.
 */

import { PostgresClient } from './postgres';
import { RedisClient } from './redis';
import { KafkaClient } from './kafka';
import { TigerBeetleClient } from './tigerbeetle';
import { MojaloopClient } from './mojaloop';
import { APISixClient } from './apisix';
import { KeycloakClient } from './keycloak';
import { OpenAppSecClient } from './openappsec';
import { PermifyClient } from './permify';
import { OpenSearchClient } from './opensearch';
import { FluvioClient } from './fluvio';
import { DaprClient } from './dapr';

export interface PlatformConfig {
  postgresUrl: string;
  redisAddr: string;
  kafkaBrokers: string[];
  tigerbeetleAddr: string;
  mojaloopUrl: string;
  apisixAdminUrl: string;
  keycloakRealmUrl: string;
  keycloakAdminUrl: string;
  keycloakClientId: string;
  keycloakClientSecret: string;
  openappsecUrl: string;
  permifyUrl: string;
  permifyTenantId: string;
  opensearchUrl: string;
  fluvioEndpoint: string;
  daprHttpPort: number;
}

export const defaultConfig: PlatformConfig = {
  postgresUrl: 'postgresql://localhost:5432/ngapp',
  redisAddr: 'localhost:6379',
  kafkaBrokers: ['localhost:9092'],
  tigerbeetleAddr: 'localhost:3000',
  mojaloopUrl: 'http://localhost:4000',
  apisixAdminUrl: 'http://localhost:9180',
  keycloakRealmUrl: 'http://localhost:8080/realms/insurance',
  keycloakAdminUrl: 'http://localhost:8080',
  keycloakClientId: 'ngapp-platform',
  keycloakClientSecret: '',
  openappsecUrl: 'http://localhost:8090',
  permifyUrl: 'http://localhost:3476',
  permifyTenantId: 'ngapp',
  opensearchUrl: 'http://localhost:9200',
  fluvioEndpoint: 'localhost:9003',
  daprHttpPort: 3500,
};

export class Platform {
  public readonly postgres: PostgresClient;
  public readonly redis: RedisClient;
  public readonly kafka: KafkaClient;
  public readonly tigerbeetle: TigerBeetleClient;
  public readonly mojaloop: MojaloopClient;
  public readonly apisix: APISixClient;
  public readonly keycloak: KeycloakClient;
  public readonly openappsec: OpenAppSecClient;
  public readonly permify: PermifyClient;
  public readonly opensearch: OpenSearchClient;
  public readonly fluvio: FluvioClient;
  public readonly dapr: DaprClient;

  constructor(config: Partial<PlatformConfig> = {}) {
    const cfg = { ...defaultConfig, ...config };
    this.postgres = new PostgresClient(cfg.postgresUrl);
    this.redis = new RedisClient(cfg.redisAddr);
    this.kafka = new KafkaClient(cfg.kafkaBrokers);
    this.tigerbeetle = new TigerBeetleClient(cfg.tigerbeetleAddr);
    this.mojaloop = new MojaloopClient(cfg.mojaloopUrl);
    this.apisix = new APISixClient(cfg.apisixAdminUrl);
    this.keycloak = new KeycloakClient(cfg.keycloakRealmUrl, cfg.keycloakClientId, cfg.keycloakClientSecret, cfg.keycloakAdminUrl);
    this.openappsec = new OpenAppSecClient(cfg.openappsecUrl);
    this.permify = new PermifyClient(cfg.permifyUrl, cfg.permifyTenantId);
    this.opensearch = new OpenSearchClient(cfg.opensearchUrl);
    this.fluvio = new FluvioClient(cfg.fluvioEndpoint);
    this.dapr = new DaprClient(cfg.daprHttpPort);
  }

  async healthCheck(): Promise<Record<string, boolean>> {
    const components: Record<string, { ping(): Promise<void> }> = {
      postgres: this.postgres,
      redis: this.redis,
      kafka: this.kafka,
      tigerbeetle: this.tigerbeetle,
      mojaloop: this.mojaloop,
      apisix: this.apisix,
      keycloak: this.keycloak,
      openappsec: this.openappsec,
      permify: this.permify,
      opensearch: this.opensearch,
      fluvio: this.fluvio,
      dapr: this.dapr,
    };

    const results: Record<string, boolean> = {};
    await Promise.all(
      Object.entries(components).map(async ([name, client]) => {
        try {
          await client.ping();
          results[name] = true;
        } catch (err) {
          console.error(`[platform] health check failed for ${name}:`, err instanceof Error ? err.message : err);
          results[name] = false;
        }
      })
    );
    return results;
  }

  async close(): Promise<void> {
    await this.redis.close();
  }
}

File: infrastructure/rust-sdk/src/platform.rs
//! Unified platform client holding all 12 infrastructure component clients.

use crate::{
    apisix::APISixClient, dapr::DaprClient, fluvio::FluvioClient, kafka::KafkaClient,
    keycloak::KeycloakClient, mojaloop::MojaloopClient, openappsec::OpenAppSecClient,
    opensearch::OpenSearchClient, permify::PermifyClient, postgres::PostgresClient,
    redis::RedisClient, tigerbeetle::TigerBeetleClient,
};
use std::collections::HashMap;

pub struct PlatformConfig {
    pub postgres_url: String,
    pub redis_addr: String,
    pub kafka_brokers: Vec<String>,
    pub tigerbeetle_addr: String,
    pub mojaloop_url: String,
    pub apisix_admin_url: String,
    pub keycloak_realm_url: String,
    pub keycloak_client_id: String,
    pub keycloak_client_secret: String,
    pub openappsec_url: String,
    pub permify_url: String,
    pub permify_tenant_id: String,
    pub opensearch_url: String,
    pub fluvio_endpoint: String,
    pub dapr_http_port: u16,
}

impl Default for PlatformConfig {
    fn default() -> Self {
        Self {
            postgres_url: "postgresql://localhost:5432/ngapp".into(),
            redis_addr: "localhost:6379".into(),
            kafka_brokers: vec!["localhost:9092".into()],
            tigerbeetle_addr: "localhost:3000".into(),
            mojaloop_url: "http://localhost:4000".into(),
            apisix_admin_url: "http://localhost:9180".into(),
            keycloak_realm_url: "http://localhost:8080/realms/insurance".into(),
            keycloak_client_id: "ngapp-platform".into(),
            keycloak_client_secret: String::new(),
            openappsec_url: "http://localhost:8090".into(),
            permify_url: "http://localhost:3476".into(),
            permify_tenant_id: "ngapp".into(),
            opensearch_url: "http://localhost:9200".into(),
            fluvio_endpoint: "localhost:9003".into(),
            dapr_http_port: 3500,
        }
    }
}

pub struct Platform {
    pub postgres: PostgresClient,
    pub redis: RedisClient,
    pub kafka: KafkaClient,
    pub tigerbeetle: TigerBeetleClient,
    pub mojaloop: MojaloopClient,
    pub apisix: APISixClient,
    pub keycloak: KeycloakClient,
    pub openappsec: OpenAppSecClient,
    pub permify: PermifyClient,
    pub opensearch: OpenSearchClient,
    pub fluvio: FluvioClient,
    pub dapr: DaprClient,
}

impl Platform {
    pub fn new(cfg: PlatformConfig) -> Self {
        Self {
            postgres: PostgresClient::new(&cfg.postgres_url),
            redis: RedisClient::new(&cfg.redis_addr),
            kafka: KafkaClient::new(cfg.kafka_brokers),
            tigerbeetle: TigerBeetleClient::new(&cfg.tigerbeetle_addr),
            mojaloop: MojaloopClient::new(&cfg.mojaloop_url),
            apisix: APISixClient::new(&cfg.apisix_admin_url),
            keycloak: KeycloakClient::new(&cfg.keycloak_realm_url, &cfg.keycloak_client_id, &cfg.keycloak_client_secret),
            openappsec: OpenAppSecClient::new(&cfg.openappsec_url),
            permify: PermifyClient::new(&cfg.permify_url, &cfg.permify_tenant_id),
            opensearch: OpenSearchClient::new(&cfg.opensearch_url),
            fluvio: FluvioClient::new(&cfg.fluvio_endpoint),
            dapr: DaprClient::new(cfg.dapr_http_port),
        }
    }

    pub async fn health_check(&self) -> HashMap<String, bool> {
        let mut results = HashMap::new();
        results.insert("postgres".into(), self.postgres.ping().await.is_ok());
        results.insert("redis".into(), self.redis.ping().await.is_ok());
        results.insert("kafka".into(), self.kafka.ping().await.is_ok());
        results.insert("tigerbeetle".into(), self.tigerbeetle.ping().await.is_ok());
        results.insert("mojaloop".into(), self.mojaloop.ping().await.is_ok());
        results.insert("apisix".into(), self.apisix.ping().await.is_ok());
        results.insert("keycloak".into(), self.keycloak.ping().await.is_ok());
        results.insert("openappsec".into(), self.openappsec.ping().await.is_ok());
        results.insert("permify".into(), self.permify.ping().await.is_ok());
        results.insert("opensearch".into(), self.opensearch.ping().await.is_ok());
        results.insert("fluvio".into(), self.fluvio.ping().await.is_ok());
        results.insert("dapr".into(), self.dapr.ping().await.is_ok());
        results
    }
}
